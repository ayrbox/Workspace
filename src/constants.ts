export interface ShiftType {
  key: string;
  label: string;
  description?: string;
}

export interface WorkspaceType {
  key: string;
  label: string;
  description?: string;
  state: 0 | 1; // working state
  isDefault?: boolean;
  color: string;
}

export const WORKSPACES: WorkspaceType[] = [
  {
    key: 'HOME',
    label: 'Home',
    description: 'Working from Home',
    state: 1,
    color: '#c8f5c8',
  },
  {
    key: 'OFFICE',
    label: '',
    description: 'Working in Office',
    state: 1,
    isDefault: true,
    color: '#fefefe',
  },
  {
    key: 'OFF',
    label: '',
    description: 'Day off of work / On Leave',
    state: 0,
    color: '#ffdead',
  },
  {
    key: 'NONWORKINGDAY',
    label: '',
    description: 'Non working day / Weekends / Bank holiday',
    state: 0,
    color: '#f7f2cd',
  },
];

export const UNKNOWN_WORKSPACE: WorkspaceType = {
  key: 'UNKNOWN',
  label: 'Not known',
  description: 'Not known / Undefined / Null',
  state: 0,
  color: '#ddd',
};

export const SHIFTS: ShiftType[] = [
  {
    key: 'AM',
    label: 'Morning',
    description: 'Morning Shift 09:00 - 12:30',
  },
  {
    key: 'PM',
    label: 'Afternoon',
    description: 'Afternoon shift 14:00 - 18:00',
  },
];

export const UNKNOWN_SHIFT: ShiftType = {
  key: 'UNKNOWN',
  label: 'Unknown Shift',
};
