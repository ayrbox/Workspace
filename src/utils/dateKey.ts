import moment from 'moment';

const dateKey = (date: Date) : string => {
  return moment(date).format('YYYYMMDD')
} 
  
export default dateKey;
