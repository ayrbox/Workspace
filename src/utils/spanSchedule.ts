import { scheduleType, shiftScheduleType } from '../types/scheduleType';
import dateKey from './dateKey';
import { WORKSPACES, ShiftType } from '../constants';
import flatten from 'lodash/flatten';

export type spanScheduleType = {
  dayKey: string;
  shift: string;
  workspace: string;
  span: number;
};

const isWeekend = (day: Date): boolean => {
  const d = day.getDay();
  return d === 6 || d === 0;
};

export const generateRandomSchedule = (days: Date[], shifts: ShiftType[]): scheduleType => {
  return days.reduce((schedule: scheduleType, day): scheduleType => {
    const dKey = dateKey(day);
    return {
      ...schedule,
      [dKey]: shifts.reduce((shifts: shiftScheduleType, shift): shiftScheduleType => {
        const rndSpace = Math.floor(Math.random() * Math.floor(3));
        const { key } = WORKSPACES[rndSpace];
        return {
          ...shifts,
          [shift.key]: isWeekend(day) ? 'NONWORKINGDAY' : key,
        };
      }, {}),
    };
  }, {});
};

export const flattenToArray = (schedule: scheduleType): spanScheduleType[] => {
  // flatten the structure
  const flattenScheduleDays = Object.entries(schedule).map(([key, value]) => ({
    dayKey: key,
    shifts: value,
  }));

  const flattenScheduleShifts = flattenScheduleDays.map(({ dayKey, shifts }) => {
    return Object.entries(shifts).map(([key, value]) => ({
      dayKey: dayKey,
      shift: key,
      workspace: value,
      span: 1,
    }));
  });

  const flatSchedule = flatten(flattenScheduleShifts);

  return flatSchedule;
};

export function spanSchedule(flatScheule: spanScheduleType[]): spanScheduleType[] {
  let current: spanScheduleType = {
    dayKey: '',
    shift: '',
    workspace: '',
    span: 0,
  };

  let cnt = 0;

  const spannedSchedule: spanScheduleType[] = [];

  for (let i = 0; i < flatScheule.length; i++) {
    if (flatScheule[i].workspace !== current.workspace) {
      if (cnt > 0) {
        // document.write(current + ' comes --> ' + cnt + ' times<br>');
        // console.log(`${current.workspace} comes ${cnt} times`);
        spannedSchedule.push({
          ...current,
          span: cnt,
        });
      }
      current = flatScheule[i];
      cnt = 1;
    } else {
      cnt++;
    }
  }
  if (cnt > 0) {
    // console.log(`${current.workspace} comes ${cnt} times`);
    spannedSchedule.push({
      ...current,
      span: cnt,
    });
  }
  return spannedSchedule;
}

export default spanSchedule;
