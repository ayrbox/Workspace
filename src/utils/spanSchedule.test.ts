import { generateRandomSchedule, flattenToArray, spanSchedule } from './spanSchedule';
import { ShiftType } from '../constants';
import dateKey from './dateKey';

const sampleDays: Date[] = [
  new Date(2019, 8, 10),
  new Date(2019, 8, 11),
  new Date(2019, 8, 12),
  new Date(2019, 8, 13),
  new Date(2019, 8, 14),
];

const sampleShifts: ShiftType[] = [
  {
    key: 'AM',
    label: 'Morning',
  },
  {
    key: 'PM',
    label: 'Evening',
  },
];

describe('generateRandomSchedule', () => {
  it('returns an object', () => {
    const randomSchedule = generateRandomSchedule(sampleDays, sampleShifts);

    expect(Object.keys(randomSchedule).length).toBe(sampleDays.length);
    expect(Object.keys(randomSchedule)).toEqual(sampleDays.map(d => dateKey(d)));
  });
});

describe('flattenToArray', () => {
  it('return flatten schedule', () => {
    const scheduleToFlatten = {
      '20190910': { AM: 'OFF', PM: 'OFFICE' },
      '20190911': { AM: 'OFFICE', PM: 'OFF' },
      '20190912': { AM: 'HOME', PM: 'OFFICE' },
      '20190913': { AM: 'OFFICE', PM: 'OFFICE' },
      '20190914': { AM: 'OFF', PM: 'OFF' },
    };

    const flat = flattenToArray(scheduleToFlatten);

    expect(flat.length).toEqual(sampleDays.length * sampleShifts.length);
    expect(flat).toMatchSnapshot();
  });
});

describe('spanSchedule', () => {
  it('returns span schedule', () => {
    const flatScheduleToSpan = [
      { dayKey: '20190910', shift: 'AM', workspace: 'HOME', span: 1 },
      { dayKey: '20190910', shift: 'PM', workspace: 'OFF', span: 1 },
      { dayKey: '20190911', shift: 'AM', workspace: 'OFF', span: 1 },
      { dayKey: '20190911', shift: 'PM', workspace: 'HOME', span: 1 },
      { dayKey: '20190912', shift: 'AM', workspace: 'HOME', span: 1 },
      { dayKey: '20190912', shift: 'PM', workspace: 'OFFICE', span: 1 },
      { dayKey: '20190913', shift: 'AM', workspace: 'OFF', span: 1 },
      { dayKey: '20190913', shift: 'PM', workspace: 'OFFICE', span: 1 },
      { dayKey: '20190914', shift: 'AM', workspace: 'OFFICE', span: 1 },
      { dayKey: '20190914', shift: 'PM', workspace: 'OFF', span: 1 },
    ];

    const spanned = spanSchedule(flatScheduleToSpan);
    expect(spanned).toMatchSnapshot();
  });
});
