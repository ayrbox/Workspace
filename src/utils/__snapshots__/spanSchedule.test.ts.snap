// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flattenToArray return flatten schedule 1`] = `
Array [
  Object {
    "dayKey": "20190910",
    "shift": "AM",
    "span": 1,
    "workspace": "OFF",
  },
  Object {
    "dayKey": "20190910",
    "shift": "PM",
    "span": 1,
    "workspace": "OFFICE",
  },
  Object {
    "dayKey": "20190911",
    "shift": "AM",
    "span": 1,
    "workspace": "OFFICE",
  },
  Object {
    "dayKey": "20190911",
    "shift": "PM",
    "span": 1,
    "workspace": "OFF",
  },
  Object {
    "dayKey": "20190912",
    "shift": "AM",
    "span": 1,
    "workspace": "HOME",
  },
  Object {
    "dayKey": "20190912",
    "shift": "PM",
    "span": 1,
    "workspace": "OFFICE",
  },
  Object {
    "dayKey": "20190913",
    "shift": "AM",
    "span": 1,
    "workspace": "OFFICE",
  },
  Object {
    "dayKey": "20190913",
    "shift": "PM",
    "span": 1,
    "workspace": "OFFICE",
  },
  Object {
    "dayKey": "20190914",
    "shift": "AM",
    "span": 1,
    "workspace": "OFF",
  },
  Object {
    "dayKey": "20190914",
    "shift": "PM",
    "span": 1,
    "workspace": "OFF",
  },
]
`;

exports[`spanSchedule returns span schedule 1`] = `
Array [
  Object {
    "dayKey": "20190910",
    "shift": "AM",
    "span": 1,
    "workspace": "HOME",
  },
  Object {
    "dayKey": "20190910",
    "shift": "PM",
    "span": 2,
    "workspace": "OFF",
  },
  Object {
    "dayKey": "20190911",
    "shift": "PM",
    "span": 2,
    "workspace": "HOME",
  },
  Object {
    "dayKey": "20190912",
    "shift": "PM",
    "span": 1,
    "workspace": "OFFICE",
  },
  Object {
    "dayKey": "20190913",
    "shift": "AM",
    "span": 1,
    "workspace": "OFF",
  },
  Object {
    "dayKey": "20190913",
    "shift": "PM",
    "span": 2,
    "workspace": "OFFICE",
  },
  Object {
    "dayKey": "20190914",
    "shift": "PM",
    "span": 1,
    "workspace": "OFF",
  },
]
`;
